### PR TITLE
Enable rollbacks in wallet layer

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#650 

# Overview

- [x] I have added an extra callback to the restoration worker to rollback on ... rollbacks.
- [x] Made sure we don't purge the very first checkpoint we create in the wallet. We might roll back up to the genesis block, so this checkpoint is needed (and currently discarded after `k` blocks)
- [x] Added a `listCheckpoints` function to the database layer to return a full list of block headers to use to find an initial intersection. The `greatestCommonBlockHeader` now works fine with sparse lists so no work to do in this area :tada:
- [x] Extended the state-machine to also yield `listCheckpoints` calls

# Comments

### CODE REVIEW

File: lib/core/test/unit/Cardano/Wallet/DB/Properties.hs [hunk 1]
```diff
@@ -761,6 +762,6 @@ newtype GenSparseCheckpointsArgs
 
 instance Arbitrary GenSparseCheckpointsArgs where
     arbitrary = do
-        k <- fromIntegral @Int <$> suchThat arbitrary (>10)
-        h <- fromIntegral @Int . getNonNegative <$> arbitrary
+        k <- (\x -> 10 + (x `mod` 1000)) <$> arbitrary
+        h <- (`mod` 100000) <$> arbitrary
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/DB/Properties.hs:766]: So, the values generated didn't allow to make the property fail after the initial checkpoint change. We need actually a wider range of values than what the `Int` generator offers us.

**Comment by KtorZ** [comment]
bors try

**Comment by KtorZ** [comment]
bors try

**Comment by KtorZ** [comment]
bors try

**Review by rvl — APPROVED** [review]
Looks good

**Comment by rvl** [comment]
Tests pass locally for me.

bors r+

**Pull request merged**